File: src/algorithm/backtrack/No22.py
from typing import List


class No22:
    def generateParenthesis(self, n: int) -> List[str]:
        
        count = 0
        curr = 0 # how many left pa left

        res = []
        def backtrack(s, count, curr):
            if count == n and curr == 0:
                res.append(s)
                return

            if count < n:
                s += "("
                curr += 1
                count += 1
                backtrack(s, count, curr)
                s = s[:-1]
                curr -= 1
                count -= 1
            
            if curr > 0:
                s += ")"
                curr -= 1
                backtrack(s, count, curr)
                s = s[:-1]
                curr += 1
        
        backtrack("", 0, 0)

        return res

File: src/algorithm/backtrack/No22.java
package src.algorithm.backtrack;

import java.util.ArrayList;
import java.util.List;

public class No22 {
    int n;
    List<String> res;
    public List<String> generateParenthesis(int n) {
        this.n = n;
        res = new ArrayList<>();
        backtrack(new StringBuilder(), 0, 0);
        return res;
    }

    private void backtrack(StringBuilder ans, int leftCount, int count) {
        if (ans.length() == 2 * n) {
            res.add(ans.toString());
        }
        if (count < n) {
            ans.append("(");
            backtrack(ans, leftCount + 1, count+1);
            ans.deleteCharAt(ans.length()-1);
        }

        if (leftCount > 0) {
            ans.append(')');
            backtrack(ans, leftCount-1, count);
            ans.deleteCharAt(ans.length()-1);
        }
    }

    public static void main(String[] args) {
        System.out.println(new No22().generateParenthesis(3));
    }
}
